<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Deniz Akdemir" />

<meta name="date" content="2025-12-21" />

<title>Advanced Configuration and Parameterization</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Advanced Configuration and
Parameterization</h1>
<h4 class="author">Deniz Akdemir</h4>
<h4 class="date">2025-12-21</h4>


<div id="TOC">
<ul>
<li><a href="#overview" id="toc-overview">Overview</a></li>
<li><a href="#understanding-output-scales-%CF%83-vs-%C5%9D" id="toc-understanding-output-scales-σ-vs-ŝ">1. Understanding Output
Scales: Σ̂ vs Ŝ</a>
<ul>
<li><a href="#why-two-outputs" id="toc-why-two-outputs">Why Two
Outputs?</a></li>
<li><a href="#mathematical-relationship" id="toc-mathematical-relationship">Mathematical Relationship</a></li>
<li><a href="#normalization-schemes" id="toc-normalization-schemes">Normalization Schemes</a></li>
<li><a href="#example-comparing-normalizations" id="toc-example-comparing-normalizations">Example: Comparing
Normalizations</a></li>
</ul></li>
<li><a href="#scale-parameter-estimation" id="toc-scale-parameter-estimation">2. Scale Parameter Estimation</a>
<ul>
<li><a href="#when-to-use-scale_method-estimate" id="toc-when-to-use-scale_method-estimate">When to Use
<code>scale_method = &quot;estimate&quot;</code></a></li>
<li><a href="#interpreting-alpha-parameters" id="toc-interpreting-alpha-parameters">Interpreting Alpha
Parameters</a></li>
<li><a href="#scale-estimation-with-grms" id="toc-scale-estimation-with-grms">Scale Estimation with GRMs</a></li>
</ul></li>
<li><a href="#initialization-strategies" id="toc-initialization-strategies">3. Initialization Strategies</a>
<ul>
<li><a href="#available-initialization-methods" id="toc-available-initialization-methods">Available Initialization
Methods</a></li>
<li><a href="#when-to-change-initialization" id="toc-when-to-change-initialization">When to Change
Initialization</a></li>
<li><a href="#example-comparing-initializations" id="toc-example-comparing-initializations">Example: Comparing
Initializations</a></li>
</ul></li>
<li><a href="#control-parameters" id="toc-control-parameters">4. Control
Parameters</a>
<ul>
<li><a href="#convergence-control" id="toc-convergence-control">Convergence Control</a></li>
<li><a href="#numerical-stability-control" id="toc-numerical-stability-control">Numerical Stability
Control</a></li>
<li><a href="#control-parameter-interaction" id="toc-control-parameter-interaction">Control Parameter
Interaction</a></li>
</ul></li>
<li><a href="#bootstrap-configuration" id="toc-bootstrap-configuration">5. Bootstrap Configuration</a>
<ul>
<li><a href="#key-parameters" id="toc-key-parameters">Key
Parameters</a></li>
<li><a href="#example-publication-quality-bootstrap" id="toc-example-publication-quality-bootstrap">Example:
Publication-Quality Bootstrap</a></li>
<li><a href="#monitoring-bootstrap-convergence" id="toc-monitoring-bootstrap-convergence">Monitoring Bootstrap
Convergence</a></li>
</ul></li>
<li><a href="#sem-supplemented-em-configuration" id="toc-sem-supplemented-em-configuration">6. SEM (Supplemented EM)
Configuration</a>
<ul>
<li><a href="#when-to-use-sem" id="toc-when-to-use-sem">When to Use
SEM</a></li>
<li><a href="#sem-control-parameters" id="toc-sem-control-parameters">SEM Control Parameters</a></li>
<li><a href="#example-sem-with-validation" id="toc-example-sem-with-validation">Example: SEM with
Validation</a></li>
<li><a href="#sem-diagnostics" id="toc-sem-diagnostics">SEM
Diagnostics</a></li>
</ul></li>
<li><a href="#summary-configuration-decision-tree" id="toc-summary-configuration-decision-tree">Summary: Configuration
Decision Tree</a></li>
<li><a href="#further-reading" id="toc-further-reading">Further
Reading</a></li>
</ul>
</div>

<div id="overview" class="section level2">
<h2>Overview</h2>
<p>This vignette provides a deep dive into <strong>CovCombR</strong>’s
advanced configuration options. It is intended for users who need
fine-grained control over the estimation process or want to understand
the mathematical details behind the dual-scale parameterization.</p>
<p><strong>Topics covered:</strong></p>
<ol style="list-style-type: decimal">
<li>Understanding output scales (Σ̂ vs Ŝ)</li>
<li>Scale parameter estimation (α parameters)</li>
<li>Initialization strategies</li>
<li>Control parameters for convergence and stability</li>
<li>Bootstrap configuration</li>
<li>SEM (Supplemented EM) configuration</li>
</ol>
<p><strong>Prerequisites:</strong> Familiarity with basic usage of
<code>fit_covcomb()</code>. See
<code>vignette(&quot;CovCombR-vignette&quot;)</code> for an introduction.</p>
<hr />
</div>
<div id="understanding-output-scales-σ-vs-ŝ" class="section level2">
<h2>1. Understanding Output Scales: Σ̂ vs Ŝ</h2>
<div id="why-two-outputs" class="section level3">
<h3>Why Two Outputs?</h3>
<p><code>fit_covcomb()</code> returns <strong>two</strong> covariance
matrices:</p>
<ul>
<li><code>$Sigma_hat</code> (accessed via <code>coef(fit)</code>): The
<strong>primary output</strong></li>
<li><code>$S_hat</code> (accessed via <code>fitted(fit)</code>): An
<strong>alternative scale presentation</strong></li>
</ul>
<p><strong>TL;DR for most users:</strong> Use <code>Sigma_hat</code> for
everything. Ignore <code>S_hat</code> unless you have a specific reason
to use it.</p>
</div>
<div id="mathematical-relationship" class="section level3">
<h3>Mathematical Relationship</h3>
<p>The model assumes each study k observes:</p>
<p><span class="math display">\[\mathbf{S}_k \sim \text{Wishart}(\nu_k,
\alpha_k \boldsymbol{\Sigma})\]</span></p>
<p>where: - <span class="math inline">\(\boldsymbol{\Sigma}\)</span> =
common underlying covariance (what we want to estimate) - <span class="math inline">\(\alpha_k\)</span> = study-specific scale
factor</p>
<p><strong>When <code>scale_method = &quot;none&quot;</code> (default):</strong> -
All <span class="math inline">\(\alpha_k = 1\)</span> (fixed) - <span class="math inline">\(\boldsymbol{\Sigma}\)</span> estimated directly -
<strong>Result:</strong> <code>Sigma_hat</code> = <code>S_hat</code>
(identical)</p>
<p><strong>When <code>scale_method = &quot;estimate&quot;</code>:</strong> - <span class="math inline">\(\alpha_k\)</span> estimated jointly with <span class="math inline">\(\boldsymbol{\Sigma}\)</span> - Two normalization
choices: geometric mean (default) or arithmetic mean -
<strong>Result:</strong> <code>Sigma_hat</code> and <code>S_hat</code>
may differ</p>
</div>
<div id="normalization-schemes" class="section level3">
<h3>Normalization Schemes</h3>
<p><strong>Geometric mean normalization (default,
<code>alpha_normalization = &quot;geometric&quot;</code>):</strong></p>
<p><span class="math display">\[\prod_{k=1}^K \alpha_k = 1 \quad
\Rightarrow \quad \text{Sigma\_hat} = \text{S\_hat}\]</span></p>
<p>The α parameters are normalized so their geometric mean is 1. This
makes Σ̂ and Ŝ identical.</p>
<p><strong>Arithmetic mean normalization
(<code>alpha_normalization = &quot;arithmetic&quot;</code>):</strong></p>
<p><span class="math display">\[\frac{1}{K} \sum_{k=1}^K \alpha_k =
1\]</span></p>
<p>The α parameters are normalized so their arithmetic mean is 1. In
this case:</p>
<p><span class="math display">\[\text{S\_hat} = \left(\frac{1}{K}
\sum_{k=1}^K \alpha_k\right) \times \text{Sigma\_hat}\]</span></p>
<p><strong>Which to use?</strong> - <strong>For reporting
results:</strong> Always use <code>Sigma_hat</code> (via
<code>coef(fit)</code>) - <strong>For method comparison:</strong>
<code>S_hat</code> can provide an alternative view when using arithmetic
normalization</p>
</div>
<div id="example-comparing-normalizations" class="section level3">
<h3>Example: Comparing Normalizations</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(CovCombR)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># Simulate data with heterogeneous scales</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2025</span>)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">6</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>var_names <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;V&quot;</span>, <span class="dv">1</span><span class="sc">:</span>p)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>true_Sigma <span class="ot">&lt;-</span> <span class="fu">diag</span>(p)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">dimnames</span>(true_Sigma) <span class="ot">&lt;-</span> <span class="fu">list</span>(var_names, var_names)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co"># Study 1: standard scale (α = 1)</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>S1 <span class="ot">&lt;-</span> <span class="fu">rWishart</span>(<span class="dv">1</span>, <span class="dv">50</span>, true_Sigma[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])[,,<span class="dv">1</span>] <span class="sc">/</span> <span class="dv">50</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="fu">dimnames</span>(S1) <span class="ot">&lt;-</span> <span class="fu">list</span>(var_names[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], var_names[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co"># Study 2: inflated scale (α = 2)</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>S2 <span class="ot">&lt;-</span> <span class="fu">rWishart</span>(<span class="dv">1</span>, <span class="dv">60</span>, <span class="dv">2</span> <span class="sc">*</span> true_Sigma[<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>, <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>])[,,<span class="dv">1</span>] <span class="sc">/</span> <span class="dv">60</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="fu">dimnames</span>(S2) <span class="ot">&lt;-</span> <span class="fu">list</span>(var_names[<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>], var_names[<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="co"># Fit with geometric normalization (default)</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>fit_geom <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a>  <span class="at">S_list =</span> <span class="fu">list</span>(<span class="at">S1 =</span> S1, <span class="at">S2 =</span> S2),</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a>  <span class="at">nu =</span> <span class="fu">c</span>(<span class="at">S1 =</span> <span class="dv">50</span>, <span class="at">S2 =</span> <span class="dv">60</span>),</span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a>  <span class="at">scale_method =</span> <span class="st">&quot;estimate&quot;</span>,</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a>  <span class="at">alpha_normalization =</span> <span class="st">&quot;geometric&quot;</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a>)</span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a></span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a><span class="co"># Fit with arithmetic normalization</span></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a>fit_arith <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a>  <span class="at">S_list =</span> <span class="fu">list</span>(<span class="at">S1 =</span> S1, <span class="at">S2 =</span> S2),</span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a>  <span class="at">nu =</span> <span class="fu">c</span>(<span class="at">S1 =</span> <span class="dv">50</span>, <span class="at">S2 =</span> <span class="dv">60</span>),</span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a>  <span class="at">scale_method =</span> <span class="st">&quot;estimate&quot;</span>,</span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a>  <span class="at">alpha_normalization =</span> <span class="st">&quot;arithmetic&quot;</span></span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a>)</span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a><span class="co"># Compare</span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Geometric normalization:</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="co">#&gt; Geometric normalization:</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;  Alpha:&quot;</span>, fit_geom<span class="sc">$</span>alpha_hat, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a><span class="co">#&gt;   Alpha: 0.7416765 1.348297</span></span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;  Sigma_hat[1,1]:&quot;</span>, <span class="fu">coef</span>(fit_geom)[<span class="dv">1</span>,<span class="dv">1</span>], <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a><span class="co">#&gt;   Sigma_hat[1,1]: 1.488499</span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;  S_hat[1,1]:&quot;</span>, <span class="fu">fitted</span>(fit_geom)[<span class="dv">1</span>,<span class="dv">1</span>], <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a><span class="co">#&gt;   S_hat[1,1]: 1.488499</span></span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;  Identical?&quot;</span>, <span class="fu">all.equal</span>(<span class="fu">coef</span>(fit_geom), <span class="fu">fitted</span>(fit_geom)), <span class="st">&quot;</span><span class="sc">\n\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a><span class="co">#&gt;   Identical? TRUE</span></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Arithmetic normalization:</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a><span class="co">#&gt; Arithmetic normalization:</span></span>
<span id="cb1-48"><a href="#cb1-48" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;  Alpha:&quot;</span>, fit_arith<span class="sc">$</span>alpha_hat, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-49"><a href="#cb1-49" tabindex="-1"></a><span class="co">#&gt;   Alpha: 0.7097426 1.290257</span></span>
<span id="cb1-50"><a href="#cb1-50" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;  Sigma_hat[1,1]:&quot;</span>, <span class="fu">coef</span>(fit_arith)[<span class="dv">1</span>,<span class="dv">1</span>], <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-51"><a href="#cb1-51" tabindex="-1"></a><span class="co">#&gt;   Sigma_hat[1,1]: 1.555463</span></span>
<span id="cb1-52"><a href="#cb1-52" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;  S_hat[1,1]:&quot;</span>, <span class="fu">fitted</span>(fit_arith)[<span class="dv">1</span>,<span class="dv">1</span>], <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-53"><a href="#cb1-53" tabindex="-1"></a><span class="co">#&gt;   S_hat[1,1]: 1.555463</span></span>
<span id="cb1-54"><a href="#cb1-54" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;  Identical?&quot;</span>, <span class="fu">all.equal</span>(<span class="fu">coef</span>(fit_arith), <span class="fu">fitted</span>(fit_arith)), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-55"><a href="#cb1-55" tabindex="-1"></a><span class="co">#&gt;   Identical? TRUE</span></span></code></pre></div>
<p><strong>Key takeaway:</strong> Use geometric normalization (default)
to avoid confusion between Σ̂ and Ŝ.</p>
<hr />
</div>
</div>
<div id="scale-parameter-estimation" class="section level2">
<h2>2. Scale Parameter Estimation</h2>
<div id="when-to-use-scale_method-estimate" class="section level3">
<h3>When to Use <code>scale_method = &quot;estimate&quot;</code></h3>
<p>Estimate scale factors (<span class="math inline">\(\alpha_k\)</span>) when:</p>
<p><strong>✓ Studies use different platforms or technologies</strong> -
Example: Combining GRMs from SNP array (100K markers) and whole-genome
sequencing (10M markers) - Scale factors account for different marker
densities</p>
<p><strong>✓ Different measurement units</strong> - Example: Combining
correlation matrices (scale ≈ 1) with covariance matrices (scale
variable)</p>
<p><strong>✓ Suspected heterogeneous variance</strong> - Studies measure
the same variables but in populations with different variability</p>
<p><strong>✗ Do NOT use when:</strong> - All studies use identical
protocols - You’ve already standardized the matrices - Small sample
sizes (α estimation requires data; use
<code>scale_method = &quot;none&quot;</code> with fewer than ~50 samples per
study)</p>
</div>
<div id="interpreting-alpha-parameters" class="section level3">
<h3>Interpreting Alpha Parameters</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Use the example data from setup</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(S_list, nu, <span class="at">scale_method =</span> <span class="st">&quot;estimate&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Alpha parameters:</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; Alpha parameters:</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">print</span>(fit<span class="sc">$</span>alpha_hat)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt;    Study1    Study2 </span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; 1.0489011 0.9533787</span></span></code></pre></div>
<p><strong>Interpretation:</strong> - <strong>α_k &gt; 1:</strong> Study
k has larger variance than average - <strong>α_k &lt; 1:</strong> Study
k has smaller variance than average - <strong>α_k ≈ 1:</strong> Study k
is at the average scale</p>
<p><strong>Example:</strong> - Study 1 (α = 0.85): 15% below average
variance - Study 2 (α = 1.20): 20% above average variance - Study 3 (α =
0.95): 5% below average variance</p>
</div>
<div id="scale-estimation-with-grms" class="section level3">
<h3>Scale Estimation with GRMs</h3>
<p>Genomic relationship matrices (GRMs) from different marker panels can
have different scales due to differing numbers of markers. Scale
estimation accounts for this.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Example workflow (hypothetical - requires actual genotype data)</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co"># GRM from 50K SNP array</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>GRM_50K <span class="ot">&lt;-</span> <span class="fu">compute_grm</span>(genotypes_50K)  <span class="co"># Your GRM computation function</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co"># GRM from 500K SNP array</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>GRM_500K <span class="ot">&lt;-</span> <span class="fu">compute_grm</span>(genotypes_500K)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co"># Combine with scale estimation</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>  <span class="at">S_list =</span> <span class="fu">list</span>(<span class="at">Array50K =</span> GRM_50K, <span class="at">Array500K =</span> GRM_500K),</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>  <span class="at">nu =</span> <span class="fu">c</span>(<span class="at">Array50K =</span> <span class="dv">50000</span>, <span class="at">Array500K =</span> <span class="dv">500000</span>),</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>  <span class="at">scale_method =</span> <span class="st">&quot;estimate&quot;</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>)</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co"># Check relative scales</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="fu">print</span>(fit<span class="sc">$</span>alpha_hat)</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="co"># Expected: alpha for 500K might be larger due to finer resolution</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="initialization-strategies" class="section level2">
<h2>3. Initialization Strategies</h2>
<div id="available-initialization-methods" class="section level3">
<h3>Available Initialization Methods</h3>
<p>The <code>init_sigma</code> parameter controls how the algorithm
starts:</p>
<p><strong>1. <code>&quot;avg_padded&quot;</code> (default):</strong> - Compute
study-wise average covariance for observed blocks - Pad missing blocks
with small values (ridge parameter) - <strong>Pro:</strong> Warm start,
faster convergence - <strong>Con:</strong> Can get stuck in local optima
if initial guess is poor</p>
<p><strong>2. <code>&quot;identity&quot;</code>:</strong> - Start with identity
matrix - <strong>Pro:</strong> Unbiased start, good for difficult
problems - <strong>Con:</strong> Slower convergence (more iterations
needed)</p>
<p><strong>3. Custom matrix:</strong> - Provide your own starting value
- <strong>Use case:</strong> Re-fit with previous estimate, or use
domain knowledge</p>
</div>
<div id="when-to-change-initialization" class="section level3">
<h3>When to Change Initialization</h3>
<p><strong>Use <code>&quot;identity&quot;</code> if:</strong> - Algorithm
converges to unexpected values - Suspecting multi-modality (multiple
local optima) - High condition number warnings</p>
<p><strong>Use custom init if:</strong> - Fitting many similar datasets
(use previous fit as init) - Have strong prior information about the
covariance structure</p>
</div>
<div id="example-comparing-initializations" class="section level3">
<h3>Example: Comparing Initializations</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Difficult case: chain topology with small p</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2025</span>)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>var_names <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;V&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>true_Sigma <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">5</span>)</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="fu">dimnames</span>(true_Sigma) <span class="ot">&lt;-</span> <span class="fu">list</span>(var_names, var_names)</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>S1 <span class="ot">&lt;-</span> <span class="fu">rWishart</span>(<span class="dv">1</span>, <span class="dv">100</span>, true_Sigma[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])[,,<span class="dv">1</span>] <span class="sc">/</span> <span class="dv">100</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="fu">dimnames</span>(S1) <span class="ot">&lt;-</span> <span class="fu">list</span>(var_names[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], var_names[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>S2 <span class="ot">&lt;-</span> <span class="fu">rWishart</span>(<span class="dv">1</span>, <span class="dv">100</span>, true_Sigma[<span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>])[,,<span class="dv">1</span>] <span class="sc">/</span> <span class="dv">100</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="fu">dimnames</span>(S2) <span class="ot">&lt;-</span> <span class="fu">list</span>(var_names[<span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>], var_names[<span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>])</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co"># Default initialization</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>fit_default <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>  <span class="at">S_list =</span> <span class="fu">list</span>(<span class="at">S1 =</span> S1, <span class="at">S2 =</span> S2),</span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>  <span class="at">nu =</span> <span class="fu">c</span>(<span class="at">S1 =</span> <span class="dv">100</span>, <span class="at">S2 =</span> <span class="dv">100</span>),</span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>  <span class="at">init_sigma =</span> <span class="st">&quot;avg_padded&quot;</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a>)</span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a><span class="co"># Identity initialization</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a>fit_identity <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a>  <span class="at">S_list =</span> <span class="fu">list</span>(<span class="at">S1 =</span> S1, <span class="at">S2 =</span> S2),</span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a>  <span class="at">nu =</span> <span class="fu">c</span>(<span class="at">S1 =</span> <span class="dv">100</span>, <span class="at">S2 =</span> <span class="dv">100</span>),</span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a>  <span class="at">init_sigma =</span> <span class="st">&quot;identity&quot;</span></span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a>)</span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a></span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a><span class="co"># Compare convergence</span></span>
<span id="cb4-28"><a href="#cb4-28" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Avg padded iterations:&quot;</span>, fit_default<span class="sc">$</span>convergence<span class="sc">$</span>iterations, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb4-29"><a href="#cb4-29" tabindex="-1"></a><span class="co">#&gt; Avg padded iterations: 500</span></span>
<span id="cb4-30"><a href="#cb4-30" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Identity iterations:&quot;</span>, fit_identity<span class="sc">$</span>convergence<span class="sc">$</span>iterations, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb4-31"><a href="#cb4-31" tabindex="-1"></a><span class="co">#&gt; Identity iterations: 500</span></span>
<span id="cb4-32"><a href="#cb4-32" tabindex="-1"></a></span>
<span id="cb4-33"><a href="#cb4-33" tabindex="-1"></a><span class="co"># Compare estimates</span></span>
<span id="cb4-34"><a href="#cb4-34" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Max difference:&quot;</span>, <span class="fu">max</span>(<span class="fu">abs</span>(<span class="fu">coef</span>(fit_default) <span class="sc">-</span> <span class="fu">coef</span>(fit_identity))), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb4-35"><a href="#cb4-35" tabindex="-1"></a><span class="co">#&gt; Max difference: 0.1975073</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="control-parameters" class="section level2">
<h2>4. Control Parameters</h2>
<div id="convergence-control" class="section level3">
<h3>Convergence Control</h3>
<p><strong><code>max_iter</code> (default: 1000):</strong> - Maximum
number of EM iterations - Increase if algorithm doesn’t converge -
<strong>Typical range:</strong> 50-2000</p>
<p><strong><code>tol</code> (default: 1e-6):</strong> - Relative change
threshold for convergence - Algorithm stops when <span class="math inline">\(\|\Sigma^{(t+1)} - \Sigma^{(t)}\|_F /
\|\Sigma^{(t)}\|_F &lt;\)</span> <code>tol</code> - <strong>Looser
tolerance (1e-4):</strong> Faster, less accurate - <strong>Tighter
tolerance (1e-8):</strong> Slower, more accurate</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Quick and dirty fit</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>fit_fast <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(S_list, nu,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>                        <span class="at">control =</span> <span class="fu">list</span>(<span class="at">max_iter =</span> <span class="dv">100</span>, <span class="at">tol =</span> <span class="fl">1e-4</span>))</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># High precision fit</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>fit_precise <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(S_list, nu,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>                           <span class="at">control =</span> <span class="fu">list</span>(<span class="at">max_iter =</span> <span class="dv">2000</span>, <span class="at">tol =</span> <span class="fl">1e-8</span>))</span></code></pre></div>
</div>
<div id="numerical-stability-control" class="section level3">
<h3>Numerical Stability Control</h3>
<p><strong><code>ridge</code> (default: 1e-8):</strong> - Added to
diagonal during matrix inversions - Prevents singularity when matrices
are nearly singular - <strong>Increase (1e-6 or 1e-5)</strong> if: -
Getting “matrix is singular” errors - Very small eigenvalues in
estimates - Weak identifiability warnings - <strong>Decrease
(1e-10)</strong> for high-precision problems with strong
identifiability</p>
<p><strong><code>min_eigen</code> (default: 1e-6):</strong> - Minimum
eigenvalue enforced during positive definite projection - Eigenvalues
below this threshold are set to <code>min_eigen</code> - <strong>Larger
values (1e-4):</strong> More aggressive PD enforcement - <strong>Smaller
values (1e-8):</strong> Preserve near-zero eigenvalues</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Problematic dataset with near-singularity</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>fit_stable <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(S_list, nu,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                          <span class="at">control =</span> <span class="fu">list</span>(<span class="at">ridge =</span> <span class="fl">1e-5</span>, <span class="at">min_eigen =</span> <span class="fl">1e-4</span>))</span></code></pre></div>
</div>
<div id="control-parameter-interaction" class="section level3">
<h3>Control Parameter Interaction</h3>
<p><strong>General rules:</strong> - If <strong>convergence is
slow</strong> → relax <code>tol</code>, increase <code>max_iter</code> -
If <strong>numerical instability</strong> → increase <code>ridge</code>
and <code>min_eigen</code> - If <strong>unexpected estimates</strong> →
try <code>init_sigma = &quot;identity&quot;</code></p>
<hr />
</div>
</div>
<div id="bootstrap-configuration" class="section level2">
<h2>5. Bootstrap Configuration</h2>
<div id="key-parameters" class="section level3">
<h3>Key Parameters</h3>
<p>When using <code>se_method = &quot;bootstrap&quot;</code>, configure via
<code>control$bootstrap</code>:</p>
<p><strong><code>B</code> (default: 100):</strong> - Number of bootstrap
replicates - <strong>Recommendation:</strong> - Exploratory: B = 50 -
Publication: B = 200 - High stakes: B = 500+ -
<strong>Trade-off:</strong> Larger B = more accurate SEs, longer
runtime</p>
<p><strong><code>seed</code> (default: NULL):</strong> - Set for
reproducible results - <strong>Always set in production
code!</strong></p>
<p><strong><code>parallel</code> (default: FALSE):</strong> - Use
parallel computation (via <code>parallel</code> package) -
<strong>Caution:</strong> Currently experimental - Set
<code>cores</code> to number of CPU cores (default:
<code>detectCores() - 1</code>)</p>
<p><strong><code>retain_samples</code> (default: FALSE):</strong> -
Store all bootstrap estimates - Useful for distribution analysis,
quantile-based inference - <strong>Warning:</strong> High memory usage
for large p</p>
</div>
<div id="example-publication-quality-bootstrap" class="section level3">
<h3>Example: Publication-Quality Bootstrap</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Note: In practice use B=200 for publication; here we show workflow</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>fit_pub <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  <span class="at">S_list =</span> S_list,</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="at">nu =</span> nu,</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>  <span class="at">se_method =</span> <span class="st">&quot;bootstrap&quot;</span>,</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="at">control =</span> <span class="fu">list</span>(</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>    <span class="at">bootstrap =</span> <span class="fu">list</span>(</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>      <span class="at">B =</span> <span class="dv">200</span>,              <span class="co"># 200 replicates for publication</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>      <span class="at">seed =</span> <span class="dv">2025</span>,          <span class="co"># Reproducibility</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>      <span class="at">parallel =</span> <span class="cn">FALSE</span>,     <span class="co"># Serial execution (safer)</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>      <span class="at">retain_samples =</span> <span class="cn">TRUE</span> <span class="co"># Keep all bootstrap estimates</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>    )</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>  )</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>)</span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="co"># Access bootstrap samples</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a>boot_samples <span class="ot">&lt;-</span> fit_pub<span class="sc">$</span>bootstrap_samples  <span class="co"># Array: p x p x B</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="co"># Compute 95% confidence intervals (quantile method)</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>Sigma_hat <span class="ot">&lt;-</span> <span class="fu">coef</span>(fit_pub)</span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>CI_lower <span class="ot">&lt;-</span> <span class="fu">apply</span>(boot_samples, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, quantile, <span class="at">probs =</span> <span class="fl">0.025</span>)</span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a>CI_upper <span class="ot">&lt;-</span> <span class="fu">apply</span>(boot_samples, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, quantile, <span class="at">probs =</span> <span class="fl">0.975</span>)</span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="co"># Hypothesis test: Is Sigma[1,2] significantly different from 0?</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Estimate:&quot;</span>, Sigma_hat[<span class="dv">1</span>,<span class="dv">2</span>], <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;95% CI: [&quot;</span>, CI_lower[<span class="dv">1</span>,<span class="dv">2</span>], <span class="st">&quot;,&quot;</span>, CI_upper[<span class="dv">1</span>,<span class="dv">2</span>], <span class="st">&quot;]</span><span class="sc">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
</div>
<div id="monitoring-bootstrap-convergence" class="section level3">
<h3>Monitoring Bootstrap Convergence</h3>
<p>Check if individual bootstrap fits converge:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Bootstrap convergence monitoring (use smaller B for speed)</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(S_list, nu, <span class="at">se_method =</span> <span class="st">&quot;bootstrap&quot;</span>,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                  <span class="at">control =</span> <span class="fu">list</span>(<span class="at">bootstrap =</span> <span class="fu">list</span>(<span class="at">B =</span> <span class="dv">200</span>, <span class="at">seed =</span> <span class="dv">2025</span>)))</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co"># Check convergence status (stored internally)</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co"># If many bootstrap replicates fail, SEs may be unreliable</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co"># Increase max_iter or relax tol if this occurs</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="sem-supplemented-em-configuration" class="section level2">
<h2>6. SEM (Supplemented EM) Configuration</h2>
<div id="when-to-use-sem" class="section level3">
<h3>When to Use SEM</h3>
<p>Supplemented EM (Meng &amp; Rubin, 1991) provides standard errors
based on the observed information matrix without bootstrap
resampling.</p>
<p><strong>Use SEM if:</strong> - Large p (&gt; 100) where bootstrap is
prohibitively slow - Need fast approximate SEs for diagnostics -
<strong>Condition:</strong> Well-identified problem (good overlap, not
too sparse)</p>
<p><strong>Do NOT use SEM if:</strong> - Small/moderate p (&lt; 50) →
use bootstrap instead - Weak identifiability (high condition number) -
Safety-critical application → bootstrap is gold standard</p>
</div>
<div id="sem-control-parameters" class="section level3">
<h3>SEM Control Parameters</h3>
<p>Configure via <code>control$sem</code>:</p>
<p><strong><code>h</code> (default: 1e-6):</strong> - Finite difference
step size for numerical differentiation - Smaller h = more accurate but
less stable - <strong>Typical range:</strong> 1e-8 to 1e-5</p>
<p><strong><code>ridge</code> (default: inherits from
<code>control$ridge</code>):</strong> - Ridge parameter for rate matrix
inversion - Increase if rate matrix is singular</p>
</div>
<div id="example-sem-with-validation" class="section level3">
<h3>Example: SEM with Validation</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Large problem where bootstrap is slow (hypothetical workflow)</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2025</span>)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>var_names <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;V&quot;</span>, <span class="dv">1</span><span class="sc">:</span>p)</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co"># Create large study data (simulation code omitted for brevity)</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co"># S_list_large &lt;- create_large_study_data(p = 50, K = 3)</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co"># nu_large &lt;- c(Study1 = 500, Study2 = 600, Study3 = 550)</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co"># Fit with SEM</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>fit_sem <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>  <span class="at">S_list =</span> S_list_large,</span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>  <span class="at">nu =</span> nu_large,</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>  <span class="at">se_method =</span> <span class="st">&quot;sem&quot;</span>,</span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>  <span class="at">control =</span> <span class="fu">list</span>(</span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>    <span class="at">sem =</span> <span class="fu">list</span>(<span class="at">h =</span> <span class="fl">1e-6</span>, <span class="at">ridge =</span> <span class="fl">1e-6</span>)</span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>  )</span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a>)</span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co"># CRITICAL: Validate SEM against bootstrap on a subset</span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a>p_subset <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a>indices <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span>p_subset</span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a></span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a>S_list_subset <span class="ot">&lt;-</span> <span class="fu">lapply</span>(S_list_large, <span class="cf">function</span>(S) {</span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a>  S[indices, indices]</span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a>})</span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a>fit_boot_subset <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a>  <span class="at">S_list =</span> S_list_subset,</span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a>  <span class="at">nu =</span> nu_large,</span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a>  <span class="at">se_method =</span> <span class="st">&quot;bootstrap&quot;</span>,</span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a>  <span class="at">control =</span> <span class="fu">list</span>(<span class="at">bootstrap =</span> <span class="fu">list</span>(<span class="at">B =</span> <span class="dv">200</span>, <span class="at">seed =</span> <span class="dv">2025</span>))</span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a>)</span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a></span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a>fit_sem_subset <span class="ot">&lt;-</span> <span class="fu">fit_covcomb</span>(</span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a>  <span class="at">S_list =</span> S_list_subset,</span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a>  <span class="at">nu =</span> nu_large,</span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a>  <span class="at">se_method =</span> <span class="st">&quot;sem&quot;</span>,</span>
<span id="cb9-39"><a href="#cb9-39" tabindex="-1"></a>  <span class="at">control =</span> <span class="fu">list</span>(<span class="at">sem =</span> <span class="fu">list</span>(<span class="at">h =</span> <span class="fl">1e-6</span>))</span>
<span id="cb9-40"><a href="#cb9-40" tabindex="-1"></a>)</span>
<span id="cb9-41"><a href="#cb9-41" tabindex="-1"></a></span>
<span id="cb9-42"><a href="#cb9-42" tabindex="-1"></a><span class="co"># Compare SEs</span></span>
<span id="cb9-43"><a href="#cb9-43" tabindex="-1"></a>se_boot <span class="ot">&lt;-</span> fit_boot_subset<span class="sc">$</span>Sigma_se</span>
<span id="cb9-44"><a href="#cb9-44" tabindex="-1"></a>se_sem <span class="ot">&lt;-</span> fit_sem_subset<span class="sc">$</span>Sigma_se</span>
<span id="cb9-45"><a href="#cb9-45" tabindex="-1"></a></span>
<span id="cb9-46"><a href="#cb9-46" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Mean SE (bootstrap):&quot;</span>, <span class="fu">mean</span>(se_boot, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb9-47"><a href="#cb9-47" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Mean SE (SEM):&quot;</span>, <span class="fu">mean</span>(se_sem, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb9-48"><a href="#cb9-48" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Correlation:&quot;</span>, <span class="fu">cor</span>(<span class="fu">as.vector</span>(se_boot), <span class="fu">as.vector</span>(se_sem),</span>
<span id="cb9-49"><a href="#cb9-49" tabindex="-1"></a>                        <span class="at">use =</span> <span class="st">&quot;complete.obs&quot;</span>), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
</div>
<div id="sem-diagnostics" class="section level3">
<h3>SEM Diagnostics</h3>
<p><strong>Check condition number of rate matrix:</strong> - SEM
computes the EM rate matrix (derivative of EM operator) - If this matrix
is ill-conditioned, SEs are unreliable - Warnings will be issued
automatically</p>
<p><strong>Fallback strategy:</strong> If SEM produces warnings or
unrealistic SEs, fall back to bootstrap for final analysis.</p>
<hr />
</div>
</div>
<div id="summary-configuration-decision-tree" class="section level2">
<h2>Summary: Configuration Decision Tree</h2>
<pre><code>Do you need standard errors?
├─ NO → se_method = &quot;none&quot;, use defaults
└─ YES
   ├─ Small problem (p &lt; 50)?
   │  └─ YES → se_method = &quot;bootstrap&quot;, B = 200
   └─ Large problem (p &gt; 50)?
      └─ YES → se_method = &quot;sem&quot;, validate on subset

Are studies from different platforms/scales?
├─ YES → scale_method = &quot;estimate&quot;
└─ NO → scale_method = &quot;none&quot; (default)

Having convergence issues?
├─ Slow convergence → increase max_iter, relax tol
├─ Unexpected values → try init_sigma = &quot;identity&quot;
└─ Singularity errors → increase ridge and min_eigen

Need reproducible results?
└─ YES → set control$bootstrap$seed</code></pre>
<hr />
</div>
<div id="further-reading" class="section level2">
<h2>Further Reading</h2>
<ul>
<li><code>vignette(&quot;standard-errors&quot;)</code> - SE method selection
guide</li>
<li><code>vignette(&quot;study-design-guide&quot;)</code> - Design
recommendations</li>
<li><code>vignette(&quot;troubleshooting&quot;)</code> - Common issues and
solutions</li>
<li><code>?fit_covcomb</code> - Complete parameter documentation</li>
</ul>
<hr />
<p><strong>Questions or issues?</strong> Report at: <a href="https://github.com/dakdemir-nmdp/CovCombR/issues" class="uri">https://github.com/dakdemir-nmdp/CovCombR/issues</a></p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
